# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmFrameIter.cpp
# Commit: 62480de389ff
# Full Hash: 62480de389ffba353998de24760993b2e1531232
# Author: Dmitry Bezhetskov <dbezhetskov@igalia.com>
# Date: 2020-09-11 15:10:42
# Regressor Bug: 1639153
# File Overlap Count: 1
# Description:
#   Bug 1639153 - Part 5: Remove remaining uses of Frame::tls. r=lth. CLOSED TREE
#   
#   
#   Here we remove remaining uses of Frame::tls. There are many places where
#   we use it: in debug frames, in profiling frames, in jit activation, etc.
# ==============================================================================

diff -r f2e486f1be17 -r 62480de389ff js/src/wasm/WasmFrameIter.cpp
--- a/js/src/wasm/WasmFrameIter.cpp	Thu Aug 27 10:30:46 2020 +0000
+++ b/js/src/wasm/WasmFrameIter.cpp	Thu Aug 27 10:53:04 2020 +0000
@@ -205,9 +205,7 @@
     tls_ = ExtractCallerTlsFromFrameWithTls(prevFP);
   }
 
-  MOZ_ASSERT(code_ == &fp_->instance()->code());
-  MOZ_ASSERT(fp_->instance() == tls()->instance);
-
+  MOZ_ASSERT(code_ == &tls()->instance->code());
   lineOrBytecode_ = callsite->lineOrBytecode();
 
   MOZ_ASSERT(!done());
@@ -926,7 +924,7 @@
          (offsetInCode - codeRange->funcCheckedCallEntry()) > SetFP;
 }
 
-TlsData* js::wasm::GetNearestEffectiveTls(Frame* fp) {
+const TlsData* js::wasm::GetNearestEffectiveTls(const Frame* fp) {
   while (true) {
     if (fp->callerIsExitOrJitEntryFP()) {
       // It is a direct call from JIT.
@@ -954,6 +952,11 @@
   }
 }
 
+TlsData* js::wasm::GetNearestEffectiveTls(Frame* fp) {
+  return const_cast<TlsData*>(
+      GetNearestEffectiveTls(const_cast<const Frame*>(fp)));
+}
+
 bool js::wasm::StartUnwinding(const RegisterState& registers,
                               UnwindState* unwindState, bool* unwoundCaller) {
   // Shorthands.
@@ -1325,7 +1328,8 @@
   }
 
   MOZ_ASSERT(code_ ==
-             &Frame::fromUntaggedWasmExitFP(callerFP_)->instance()->code());
+             &GetNearestEffectiveTls(Frame::fromUntaggedWasmExitFP(callerFP_))
+                  ->instance->code());
 
   switch (codeRange_->kind()) {
     case CodeRange::Function: